Make the `useCampaigns` hook support being used with multiple instances

eason9487 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #37177 

 This PR makes the `useCampaigns` hook support being used with multiple instances.

- Include the `perPage` in the page key of campaigns in the multichannel's data store.
- Move the processing of querying the campaigns paging to the selector of the multichannel's data store.
- ~Move `total` in each campaign's paging of the multichannel's data store.~ Move the `total` from `campaigns.pages[*]` to `campaigns.meta`.
- Remove unused properties and their types: `state.campaigns.perPage` and `state.campaigns.total`

💡 Currently, the `useCampaigns` hook has only one use instance, so the problems in issue #37177 don't exist. Therefore, the **Campaigns** card on the Marketing > Overview page should work as well as before.

💡 This PR is based on #41180 to include the relevant code changes.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### 📌 By test cases

1. View the `useCampaigns.test.ts` file to determine if it fulfills the test purpose
2. Run `cd plugins/woocommerce-admin`
3. Run `git checkout b1b8539a6b78ad4353a03ab9a6092161034ec824` and `npm run turbo:test -- useCampaigns`
   - Check if the added tests can catch the two problems in [#37177](https://github.com/woocommerce/woocommerce/issues/37177)
   - These last two should be failed:
      ![image](https://github.com/woocommerce/woocommerce/assets/17420811/38e2eb28-cff9-4fde-89ce-ec3a7cdedd61)
4. Run `git checkout e2651bf3b921995bd08caa3297eb9642c9a94fbf` and `npm run turbo:test -- useCampaigns`
   - Check if the "**should update data accordingly once the registered channels are updated**" case is turned to pass
   - The last one should be still failed
5. Run `git checkout tweak/37177-use-campaigns-hook-multiple-instances` and `npm run turbo:test -- useCampaigns`
   - Check if the "**should be able to use different arguments for different instances at the same time**" case is turned to pass
   - All test cases should pass as well
      ![image](https://github.com/woocommerce/woocommerce/assets/17420811/b7949fd3-566d-49a9-88dd-0af58ff0e61d)

#### 📌 By web page

1. Install [Google Listings & Ads](https://github.com/woocommerce/google-listings-and-ads) and finish the extension onboarding and ads onboarding
2. Modify the line of [GLAChannel.php#L75](https://github.com/woocommerce/google-listings-and-ads/blob/72543805c6a24191c33c5cd0177d66b07537b18b/src/MultichannelMarketing/GLAChannel.php#L75) to the following:
   ```php
   if ( apply_filters( 'woocommerce_gla_enable_mcm', true ) === true ) {
   ```
3. If the number of campaigns is less than 6, modify the `get_campaigns` function in [GLAChannel.php#L178-L206](https://github.com/woocommerce/google-listings-and-ads/blob/72543805c6a24191c33c5cd0177d66b07537b18b/src/MultichannelMarketing/GLAChannel.php#L178-L206), so that it returns some sample marketing campaigns:
   ```php
   public function get_campaigns(): array {
   	return [
   		new MarketingCampaign(
   			'1',
   			$this->campaign_types['google-ads'],
   			'Campaign 1',
   			admin_url( 'admin.php?page=wc-admin&path=/google/settings&subpath=/campaigns/edit' ),
   			new Price( '19.99', 'USD' ),
   		),
   		new MarketingCampaign(
   			'2',
   			$this->campaign_types['google-ads'],
   			'Campaign 2',
   			admin_url( 'admin.php?page=wc-admin&path=/google/settings&subpath=/campaigns/edit' ),
   			new Price( '20', 'USD' ),
   		),
   		new MarketingCampaign(
   			'3',
   			$this->campaign_types['google-ads'],
   			'Campaign 3',
   			admin_url( 'admin.php?page=wc-admin&path=/google/settings&subpath=/campaigns/edit' ),
   			new Price( '30', 'USD' ),
   		),
   		new MarketingCampaign(
   			'4',
   			$this->campaign_types['google-ads'],
   			'Campaign 4',
   			admin_url( 'admin.php?page=wc-admin&path=/google/settings&subpath=/campaigns/edit' ),
   			new Price( '400', 'USD' ),
   		),
   		new MarketingCampaign(
   			'5',
   			$this->campaign_types['google-ads'],
   			'Campaign 5',
   			admin_url( 'admin.php?page=wc-admin&path=/google/settings&subpath=/campaigns/edit' ),
   			new Price( '5,000', 'USD' ),
   		),
   		new MarketingCampaign(
   			'6',
   			$this->campaign_types['google-ads'],
   			'Campaign 6',
   			admin_url( 'admin.php?page=wc-admin&path=/google/settings&subpath=/campaigns/edit' ),
   			new Price( '6,888.5', 'USD' ),
   		),
   	];
   }
   ```
4. Go to the Marketing > Overview admin page
5. View if the **Campaigns** card works as well as before

https://github.com/woocommerce/woocommerce/assets/17420811/0ce7b90e-3b3a-4190-ae9c-63af6533bbff

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

eason9487 committed:
Add tests for the `useCampaigns` hook in the woocommerce-admin marketing.

eason9487 committed:
Add `channels` to the deps of `useSelect` in the `useCampaigns` hook to avoid outdated hook states.

eason9487 committed:
Make the `useCampaigns` hook support being used with multiple instances.

- Include the `perPage` in the page key of campaigns in the multichannel's data store.
- Move the processing of querying the campaigns paging to the selector of the multichannel's data store.
- Move `total` in each campaign's paging of the multichannel's data store.
- Remove unused properties and their types: `state.campaigns.perPage` and `state.campaigns.total`

eason9487 committed:
Add changelog

ecgan commented on plugins/woocommerce-admin/client/marketing/data-multichannel/reducer.ts:
```diff
@@ -21,9 +21,7 @@ const initialState = {
-		perPage: undefined,
-		pages: undefined,
-		total: undefined,
+		pages: {},
```
❓ If there is only one child node `pages` here, I think we should remove it, so that we can just straight away do `campaigns[key]` instead of `campaigns.pages[key]`.

eason9487 replied:
After the change related to https://github.com/woocommerce/woocommerce/pull/41211#discussion_r1411860547, the `pages` is no longer the only child of `campaigns`.

ecgan commented on plugins/woocommerce-admin/client/marketing/data-multichannel/reducer.ts:
```diff
@@ -66,21 +64,23 @@ export const reducer: Reducer< State, Action > = (
 
 		case TYPES.RECEIVE_CAMPAIGNS:
+			const { meta } = action;
+			const key = `${ meta.page }-${ meta.perPage }`;
```
❔ 💅 This is kinda personal opinion: I would prefer not to use `-` as a data separator because `-` is often used in slugs and in URLs, and may cause confusion. I would prefer something else like `|` as data separator in a string. 

Also, I have a thought to make the key more readable, e.g. "page:1|perPage:10". I think that is helpful when we look at redux dev tools and when troubleshooting. But on the other hand, this may also be considered too verbose.

For reference, here's an example screenshot from redux dev tool, showing the current structure:

<img width="348" alt="image" src="https://github.com/woocommerce/woocommerce/assets/417342/87088a32-77c1-4a95-91a0-0fd17d6a2253">

eason9487 replied:
> This is kinda personal opinion: I would prefer not to use `-` as a data separator because `-` is often used in slugs and in URLs, and may cause confusion. I would prefer something else like `|` as data separator in a string.

Globally searching the codebase of this repo with ``/const \w+ = (['"`]).+?\|.+?\1/i`` and ``/\+ (['"`])\|\1 \+/`` regex patterns only got two instances:

1. https://github.com/woocommerce/woocommerce/blob/6b77f71ea7312a18ac35047cce38ffa5574eece4/.github/actions/tests/slack-summary-on-release/slack-blocks/scripts/create-result-block/utils/set-element-text.js#L5
2. https://github.com/woocommerce/woocommerce/blob/6b77f71ea7312a18ac35047cce38ffa5574eece4/plugins/woocommerce/client/legacy/js/admin/api-keys.js#L88

Searching with ``const \w*(id|code|key) = (['"`]).+?-.+?\2`` and ``/\+ (['"`])-\1 \+/`` got dozens of instances:

![image](https://github.com/woocommerce/woocommerce/assets/17420811/40a83b78-2830-4448-bd8c-c1971e9eb65d)

It's certainly commonly used for slugs or URLs, and quite often for IDs or keys as well. Also, I don't get the point of reading ``const key = `${ meta.page }-${ meta.perPage }`;`` would be confused with a slug or a URL.

If a separator does not appear in the value range of `page` and `perPage`, is not an unrecognizable character, and does not violate convention, then I think it's reasonable.

> Also, I have a thought to make the key more readable, e.g. "page:1|perPage:10". I think that is helpful when we look at redux dev tools and when troubleshooting. But on the other hand, this may also be considered too verbose.

Since there are only two parameters, I don't think it's complicated enough to require a highly readable key.

ecgan replied:
> It's certainly commonly used for slugs or URLs, and quite often for IDs or keys as well. Also, I don't get the point of reading `` const key = `${ meta.page }-${ meta.perPage }`; `` would be confused with a slug or a URL.

I guess I should have illustrated it with a scenario. 😄 

Imagine an example scenario we see a string "hello-world-1-2", in a URL path, query string, key ID etc. At first glance, we may think it is _one data piece_ of string. However, if we dig into the code, we may find that it is actually constructed from _multiple pieces of data_, for example `${ stringHelloWorld1 }-${ number2}`. If we use another character to construct the string like `${ stringHelloWorld1 }|${ number2}`, it would produce "hello-world-1|2", which would indicate that it is more likely to be constructed from multiple data pieces.

In summary, I guess what I want to convey is that: because "-" commonly appears _within one string data piece_, I would prefer not to use it as a data separator to join _multiple data pieces_. If we do that, it makes it a bit more difficult to tell from the final string whether it comes from one string data piece or it is generated from multiple data pieces. And it may be difficult to deconstruct the final string back into the individual data pieces.

> If a separator does not appear in the value range of `page` and `perPage`, is not an unrecognizable character, and does not violate convention, then I think it's reasonable.

Yes, for our case here, I think it is okay and reasonable. I was just thinking about more complicated scenarios, and we may want to be consistent across scenarios.

We can stick with what we have here. 🙂 👌

eason9487 replied:
Thanks for the explanation!

> Imagine an example scenario we see a string "hello-world-1-2", in a URL path, query string, key ID etc. At first glance, we may think it is _one data piece_ of string. However, if we dig into the code, we may find that it is actually constructed from _multiple pieces of data_, for example `${ stringHelloWorld1 }-${ number2}`.

I would say this scenario is not analogous to the use case here. If "-" is possible in the value range of variables to be concatenated, then it should not be used as a concatenation character in the first place. Moreover, perhaps the premise that it is considered as a "data (piece)" to be discussed is not quite accurate either.

The intent of this code should be clear: it wants to generate a non-colliding data key to be concatenated only and doesn't need to be separated back for somewhat uses.

> In summary, I guess what I want to convey is that: because "-" commonly appears _within one string data piece_, I would prefer not to use it as a data separator to join _multiple data pieces_. If we do that, it makes it a bit more difficult to tell from the final string whether it comes from one string data piece or it is generated from multiple data pieces.

I would suggest not just looking at the final string to determine/infer how it came to be, instead, exactly find where it came from. It's like saying that when seeing a string "1|5", a dev still has to find out where it came from in order to truly determine how it came to be, and it doesn't make the whole thing any easier or harder.

> And it may be difficult to deconstruct the final string back into the individual data pieces.

If the character meets the principles mentioned at the end of [this comment](https://github.com/woocommerce/woocommerce/pull/41211#discussion_r1411813274), I believe it's not that difficult to separate it.

> I was just thinking about more complicated scenarios, and we may want to be consistent across scenarios.

To discuss consistency across (more complicated) scenarios, I guess [Discussions](https://github.com/woocommerce/woocommerce/discussions) would be a better place.

ecgan commented on plugins/woocommerce/changelog/tweak-37177-use-campaigns-hook-multiple-instances:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: tweak
```
❔ I was thinking this is more of a "Dev" type, because it is code / development related and doesn't have any user-facing changes. I guess it is a "dev-related code tweak", so maybe "tweak" is okay too. 😂

eason9487 replied:
This PR did change a few codes used in the Admin Marketing page rather than only adding some test cases. It seems to be closer to the tweak type.

ecgan replied:
I was thinking that "tweak" refers to very minor changes, e.g. fixing typo, fixing grammar or sentence mistakes, adjusting some pixels in an image or layout etc. Very minor user-facing changes. Because this PR are about development code changes and does not have visible user-facing changes, I thought it would be a "dev" type.

But I guess "tweak" is okay too, in the sense of "dev-related code tweak".

ecgan commented on plugins/woocommerce-admin/client/marketing/data-multichannel/types.ts:
```diff
@@ -63,13 +63,12 @@ export type Campaign = {
 
 export type CampaignsPage = {
 	data?: Array< Campaign >;
+	total?: number;
```
❓ / 🚧  I'm thinking we shouldn't move `total` into `CampaignsPage`, because this means we may have multiple different totals across pages. We should only have one latest updated `total` that is outside of `pages`.

Consider this scenario: 

1. Initially, we have 6 campaigns in the database, and we query page 1 with perPage 5. At this point, we should see "Page 1 of **2**" in the UI (because the `total` in page 1 is 6).  
2. Then, before we go to page 2, new campaigns are created, and there are now a total of 18 campaigns in the database. 
3. We go to page 2. The `total` is now 18, and we would see "Page 2 of **4**" in the UI (because the `total` in page 2 is 18). 
4. We go back to page 2. We would see "Page 1 of **2**" again. 

From the above scenario, we see that the total number of pages changed from 2 to 4, and to 2 again, which can be confusing or can be seen as a bug.

If we have one `total` that is always latest and updated, then I think we would not have this issue.

eason9487 replied:
This scenario describes a frontend cache synchronization problem across multiple browser tab instances. Even if changing `total` to have only one data and always keeping it up-to-date will only make the `total` value correct. However, the paging results may still be incorrect.

Continuing the scenario:

- For example, if sorting by creation time descending, the result of the first page is different after the total number of campaigns has increased from 6 to 18.
- For another example, if a campaign's budget is changed in a separate browser tab, and then back to the browser tab viewing the campaigns card, the cached data store still keeps the outdated data no matter switching the campaigns card back and forth.

Synchronizing frontend caching across multiple browser tabs is not usually a default concern, and I believe this "concern" exists everywhere in Woo. Therefore, I don't think this is a sound reason to change the `total` to a single instance.

Edited: I replied the above based on the multiple browser tabs, but I guess the main comment might based on the same tab. In this case, the root cause of the issue is not the same but still similar and also leads to the same issues in the scenario. And then the question would be, should an extension invalidate a Woo's data store? Probably not.

The reason for moving it is to make the reference of the values returned by `getCampaigns` selector stable. However, in terms of the design of data structure, it's a good practice to keep it as a single instance. Changed in https://github.com/woocommerce/woocommerce/pull/41211/commits/ab9127328548a43a5c8897d378866d05436ea11b.

ecgan replied:
> in terms of the design of data structure

Ahh yes, you say it better than I do in my scenario! 😄 Thank you.

ecgan commented on plugins/woocommerce-admin/client/marketing/hooks/test/useCampaigns.test.ts:
```diff
@@ -0,0 +1,230 @@
+import { Campaign as APICampaign } from '../../data-multichannel/types';
+import '../../data-multichannel'; // To ensure the store is registered
+
+type Channel = Pick< RegisteredChannel, 'slug' | 'title' | 'icon' >;
```
❔ Minor code enhancement: I'm thinking we don't need to have this custom `Channel` type. Instead, we can use a default channel data object and override it when we use it, for example:

```js
const defaultChannelData: RegisteredChannel = {
	slug: 'extension-foo',
	title: 'Extension Foo',
	icon: 'https://test/foo.png',
	description: '',
	isSetupCompleted: true,
	setupUrl: '',
	manageUrl: '',
	syncStatus: 'synced',
	issueType: 'none',
	issueText: '',
};

// then use it like this:
mockRegisteredChannels( {
	...defaultChannelData,
	slug: 'extension-foo',
	title: 'Extension Foo',
	icon: 'https://test/foo.png',
} );
```

With this approach, we wouldn't need to have the `channels.map` call in the `mockRegisteredChannels` function too.

eason9487 replied:
Comparing the single line of the type definition to the given example, this will require more lines of code and looks redundant when calling `mockRegisteredChannels`. I lean toward keeping it as it is.

ecgan commented on plugins/woocommerce-admin/client/marketing/hooks/test/useCampaigns.test.ts:
```diff
@@ -0,0 +1,230 @@
+		const page = Number( params.get( 'page' ) );
+		const perPage = Number( params.get( 'per_page' ) );
+
+		if ( ! Number.isInteger( page ) || ! Number.isInteger( perPage ) ) {
+			return Promise.reject(
+				new Response( '{"message": "invalid query"}', { status: 400 } )
+			);
+		}
+
+		const length = Math.min( perPage, total - ( page - 1 ) * perPage );
+
+		const campaigns: Array< APICampaign > = Array.from( { length } ).map(
+			( _, index ) => {
+				const id = `${ page }_${ index + 1 }`;
+				return {
+					id,
+					channel: 'extension-foo',
+					title: `Campaign ${ id }`,
+					manage_url: `https://test/extension-foo?path=setup&id=${ id }`,
+					cost: {
+						value: ( ( page * perPage + index ) * 0.25 ).toString(),
+						currency: 'USD',
+					},
+				};
+			}
+		);
+
+		// For testing fallbacks when data fields are not available
+		if ( campaigns[ 2 ] ) {
+			campaigns[ 2 ].channel = 'intentional-mismatch-channel';
+			campaigns[ 2 ].cost = null;
+		}
+
+		return Promise.resolve(
+			new Response( JSON.stringify( campaigns ), {
+				headers: new Headers( { 'x-wp-total': total.toString() } ),
+			} )
+		);
+	} )
+);
```
❔ I have some reservations / hesitance about mocking "@wordpress/api-fetch" here, because this means that we are also indirectly testing wp-data redux things here, i.e. the selector, resolver, reducer etc, in this `useCampaigns.test.js` file.

My thinking is that `useCampaigns.test.js` should be unit test on `useCampaigns`, and we would be mocking `useSelect`. But what we have here is more like integration test with wp-data, which is fine too, but just a different expectation. 😄 

So I think it's fine to keep it this way. Just want to express my expectation and opinion. If you can share your experience or thinking here, that would be great for us to learn.

eason9487 replied:
Ideally, it's proper to mock from the layer of `getCampaigns` selector rather than `@wordpress/api-fetch`. However, the root cause of this issue appears to be the mechanism of `@wordpress/data` and how the selectors and reducer of the data store access the states.

If it was mocking the layer of `getCampaigns` selector, the following test cases would not effectively validate the issue this PR is intended to solve, nor would it prevent similar issues from occurring in the future:
- https://github.com/woocommerce/woocommerce/blob/ab9127328548a43a5c8897d378866d05436ea11b/plugins/woocommerce-admin/client/marketing/hooks/test/useCampaigns.test.ts#L183
- https://github.com/woocommerce/woocommerce/blob/ab9127328548a43a5c8897d378866d05436ea11b/plugins/woocommerce-admin/client/marketing/hooks/test/useCampaigns.test.ts#L209

ecgan replied:
Got it. Thank you for the explanation! 🙂 👍

ecgan commented on plugins/woocommerce-admin/client/marketing/data-multichannel/selectors.ts:
```diff
@@ -13,9 +13,14 @@ export const getRecommendedChannels = ( state: State ) => {
-export const getCampaigns = ( state: State ) => {
-	return state.campaigns;
+export const getCampaigns = ( state: State, page: number, perPage: number ) => {
+	const key = `${ page }-${ perPage }`;
```
❔ Related to the other comment about using `-` as data separator in key: it might be good to refactor this key construction into a shared function that we can use in selector and reducer.

eason9487 replied:
Replied in https://github.com/woocommerce/woocommerce/pull/41211#discussion_r1411813274.

ecgan reviewed (commented):
@eason9487 , thanks so much for this PR! 🙏 

I tested based on the test instructions, things work as expected. 👍 

I'm leaving a few comments here. Important ones are marked with ❓ / 🚧 . 

Requesting for your comments on them before approval.

eason9487 committed:
Multichannel data store - Move the `total` from `campaigns.pages[*]` to `campaigns.meta` to have only one total value.

Related to:
- https://github.com/woocommerce/woocommerce/pull/41211#discussion_r1409634640
- https://github.com/woocommerce/woocommerce/pull/41211#discussion_r1409590559

eason9487 commented:
Hi @ecgan, thanks for the review!

I have addressed/replied to all the review comments. It's ready for the new round of code reviews. Could you help with it?

ecgan commented on plugins/woocommerce-admin/client/marketing/data-multichannel/selectors.ts:
```diff
@@ -23,6 +23,15 @@ export const getCampaigns = ( state: State, page: number, perPage: number ) => {
+export const getCampaignsMeta = ( state: State ) => {
+	return state.campaigns.meta;
+};
+
```
❓ `getCampaignsMeta` has a "dependency" on `getCampaigns`, but it is not immediately obvious here.

Imagine a scenario: someone uses `getCampaignsMeta` with `useSelect` directly in their component just to show the total number of campaigns, without using `useCampaigns` hook or `getCampaigns` selector. The `getCampaigns` resolver does not get called, so the `meta.total` would be `undefined`.

We can add a description into the JSDoc comment, e.g. "You should use `getCampaigns` or `useCampaigns` before using `getCampaignsMeta` first." But this feels like a patch to the problem; the problem would still be there.

I think a better way is to return the `meta` / `meta.total` in the `getCampaigns` selector, so that when we call `getCampaigns`, we would get the campaigns page and also the total in one selector, and the resolver would also get called.

eason9487 replied:
Maybe I was too focused on maintaining the stability of returning references.

> I think a better way is to return the `meta` / `meta.total` in the `getCampaigns` selector, so that when we call `getCampaigns`, we would get the campaigns page and also the total in one selector, and the resolver would also get called.

Considering that `getCampaigns` doesn't actually have rendering performance issues, and that even `useCampaigns` itself doesn't consider the stability of returning references, it should be fine for now.

if necessary in the future, wrapping `getCampaigns` with `rememo` or referring to something like [this resolver](https://github.com/woocommerce/woocommerce/blob/54b980c2d2512ff37c5fee3c869e325a8dce31f3/packages/js/data/src/crud/resolvers.ts#L107-L146) should make references stable.

Revised in https://github.com/woocommerce/woocommerce/pull/41211/commits/17c624ad5cbcea5671b3364da44eb1b9c40ff204.

ecgan replied:
> Considering that `getCampaigns` doesn't actually have rendering performance issues, and that even `useCampaigns` itself doesn't consider the stability of returning references, it should be fine for now.
> 
> if necessary in the future, wrapping `getCampaigns` with `rememo` or referring to something like [this resolver](https://github.com/woocommerce/woocommerce/blob/54b980c2d2512ff37c5fee3c869e325a8dce31f3/packages/js/data/src/crud/resolvers.ts#L107-L146) should make references stable.

@eason9487 , could you elaborate more on the above? Is there a potential stability issue with `useCampaigns`? How would it become unstable? 🤔

ecgan replied:
I did some digging and I think I get what you mean now. For example, in `useCampaigns`, we have this line:

```js
data: campaignsPage?.data?.map( convert ),
```

Even though the campaigns data in the data store did not change, because we are using `.map( convert )`, it returns a new array with the same data every time, so the reference is not stable and it causes unnecessary re-renders.

And I think the same issue happened for the `useCampaigns` code before this PR, where we were returning a new object for the `meta` field:

```js
				meta: {
					total: campaignsState.total,
				},
```

In this PR, because now we are returning the `meta` field directly from the data store without creating a new object, so we do not have this issue with the `meta` field.

Is there anything else that I'm missing? 🤔 

Reference: https://jardasn.blog/2021/01/22/some-best-practices-for-using-useselect-from-wordpress-data/

eason9487 replied:
Yes, it's recommended to keep the references shallowly equal to the previous one when using `useSelect`, but the `data` property still doesn't meet that after this PR.

For the `getCampaigns` selector, it shares the same concept of redux that avoids creating derived values that are new references. So there are a number of WP and Woo practices that use a memoization library, which is `rememo`, to achieve this.

ecgan reviewed (changes requested):
@eason9487 , thanks for the change and the explanations. 🙏 

I looked into your new code change, and I think there is an issue (marked with❓) with the new selector `getCampaignsMeta` which we may need to change.

eason9487 committed:
Multichannel data store - Return the pagination meta of campaigns along with the `getCampaigns` selector.

Address: https://github.com/woocommerce/woocommerce/pull/41211#discussion_r1412266127

eason9487 commented:
Hi @ecgan, thanks for your time. I have addressed the concern in https://github.com/woocommerce/woocommerce/pull/41211#discussion_r1412266127 by https://github.com/woocommerce/woocommerce/pull/41211/commits/17c624ad5cbcea5671b3364da44eb1b9c40ff204. Could you take a look?

ecgan commented on plugins/woocommerce-admin/client/marketing/data-multichannel/reducer.ts:
```diff
@@ -21,9 +21,10 @@ const initialState = {
-		pages: undefined,
-		total: undefined,
+		pages: {},
+		meta: {
+			total: undefined,
+		},
```
❔ This just came to me: I think the `meta` object is not needed, and we can actually just do `state.campaigns.total` instead of `state.campaigns.meta.total`, right?

But on the other hand, I think having `meta` has its own merits too, e.g. it stores things about the request we made, and the `total` value came from the response header, and perhaps we may store other data under `meta` here in the future etc.

ecgan replied:
Adding one more note: I think it may be good to keep it, to match with the `meta` field in the `useCampaigns` returned value.

eason9487 replied:
https://github.com/woocommerce/woocommerce/pull/41211#discussion_r1414144200 is the reason I added the `meta` object.

ecgan reviewed (approved):
Tested, works as expected, LGTM! 👍 Thanks @eason9487 !

I left just one small minor comment about `meta` below. We can change it or keep it as is.
